service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java

```diff
@@ -94,99 +93,90 @@ public class EncounterComponentsMapper {
+            .orElse(StringUtils.EMPTY);
+    }
 
-        return topics.stream()
+    private String mapListResourceToComponent(ListResource listResource) {
+        return listResource.getEntry()
+            .stream()
+            .map(EncounterComponentsMapper::getReferenceElement)
+            .map(this::getRequiredResource)
+            .map(ListResource.class::cast)
             .map(this::mapTopicListToComponent)
             .collect(Collectors.joining());
     }
 
     private String mapTopicListToComponent(ListResource topicList) {
-
         if (!CodeableConceptMappingUtils.hasCode(topicList.getCode(), List.of(TOPIC_LIST_CODE))) {
             throw new EhrMapperException(String.format("Unexpected list %s referenced in Consultation, expected list to be coded as "
                 + "Topic (EHR)", topicList.getId()));
         }
 
-        String components = mapTopicListComponents(topicList);
-
-        if (StringUtils.isAllEmpty(components)) {
-            return components;
-        }
-
-        String effectiveTime = prepareEffectiveTime(topicList);
-        String availabilityTime = prepareAvailabilityTime(topicList);
-
-        var params = CompoundStatementParameters.builder()
-            .nested(false)
-            .id(messageContext.getIdMapper().getOrNew(ResourceType.List, topicList.getIdElement()))
-            .classCode(TOPIC.getCode())
-            .compoundStatementCode(prepareCdForTopic(topicList))
-            .statusCode(COMPLETE_CODE)
-            .effectiveTime(effectiveTime)
-            .availabilityTime(availabilityTime)
-            .components(components)
-            .build();
-
-        return TemplateUtils.fillTemplate(COMPOUND_STATEMENT_TEMPLATE, params);
+        return buildCompoundStatement(
+            topicList,
+            TOPIC.getCode(),
+            prepareCdForTopic(topicList),
+            false,
+            mapTopicListComponents(topicList)
+        );
     }
 
     private String mapTopicListComponents(ListResource topicList) {
+        return String.join(
+            StringUtils.EMPTY,
+            mapCategorizedResources(topicList),
+            mapContainedResources(topicList),
+            mapUncategorizedResources(topicList)
+        );
+    }
 
-        String categories = topicList.getEntry().stream()
-            .map(entry -> entry.getItem().getReferenceElement())
-            .filter(reference -> reference.getValue().matches(LIST_REFERENCE_PATTERN))
-            .map(reference -> messageContext
-                .getInputBundleHolder()
-                .getRequiredResource(reference))
-            .filter(resource -> ResourceType.List.equals(resource.getResourceType()))
+    private String mapCategorizedResources(ListResource topicList) {
+        return topicList.getEntry().stream()
+            .map(EncounterComponentsMapper::getReferenceElement)
+            .filter(reference -> reference
+                .getValue()
+                .matches(LIST_REFERENCE_PATTERN)
```

Comment:
Also, our approach to identifying whether something is a LIST or not is janky, because it assumes that the resource isn't of the `contained` form. HOWEVER there is nothing stopping a GP System from providing `List (Heading)` as a contained reference. This defect is separate to this original ticket, but maybe you could a Jira defect for it. 🤣